### PR TITLE
Add componentName in modalDismissed event.

### PR DESCRIPTION
**Pull request by jinshin1013**
### The problem
This PR includes:
- Adding `componentName` to the `ModalDismissed` event.
- Updating the documentation on `ModalDismissed` event.

### CODE REVIEW

**Comment by guyca** [comment]
hmm
I'm not sure about this PR, could you please elaborate on why you think it's needed?

**Comment by jinshin1013** [comment]
Sorry I should have elaborated on the reasoning behind this PR. 

So to begin, `componentName` is currently being emitted to some of the navigation event listeners such as `registerComponentDidAppearListener` and `registerComponentDidDisappearListener`. And I've been personally using these listeners to log screen `appeared / disappeared` event to the Analytic tools. 

However we found the necessity in being able to specify `modalDismissed` event. We could get away with `registerComponentDidDisappearListener` but it does not specify whether it was a normal `pop` or `modal dismissal` as we may use one registered screen as 'normal' screen and a modal. 

What's your main concern about including `componentName` in the `modalDismissed` event or do you think it is redundant?

**Comment by guyca** [comment]
@jinshin1013 Thanks for the explanation. The reason we didn't include componentName in the first place is because not every layout has a componentName. External components for example don't  have it. Moreover, it can be inferred from `componentDidAppear` event.

Lets implement this on Android as well so we can merge the PR. Relevant classes:
1. https://github.com/wix/react-native-navigation/blob/master/lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/modal/ModalStack.java#L66
2. https://github.com/wix/react-native-navigation/blob/master/lib/android/app/src/main/java/com/reactnativenavigation/react/events/EventEmitter.java#L72

And you'll probably want to add `public abstract String getCurrentComponentName();` to `ViewController.java`
which will return the `componentName` from `ComponentViewController` and

And in `ParentViewController` add something like
```java
@Override
    public String getCurrentComponentName() {
        return getCurrentChild().getCurrentComponentName();
    }
```

**Comment by jinshin1013** [comment]
@guyca Yeah that makes sense!

> Moreover, it can be inferred from componentDidAppear event. 

Yeah it was more about being able to specify `modal dismissed` which cannot be inferred from `componentDidDisappear` unless you register the component with Modal specific name. 

I'll try to implement this on the Android as soon as I get some free time but thanks for the guidance!

File: lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/TitleBarButtonController.java [hunk 1]
```diff
@@ -87,6 +87,11 @@ public void sendOnNavigationButtonPressed(String buttonId) {
         getView().sendOnNavigationButtonPressed(buttonId);
     }
 
+    @Override
+    public String getCurrentComponentName() {
```
Comment (jinshin1013) [inline lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/TitleBarButtonController.java]: It was complaining that `getCurrentComponetName` needs to be implemented. Should this return its actual componentName?

File: lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/TitleBarReactViewController.java [hunk 2]
```diff
@@ -44,6 +44,11 @@ public void sendOnNavigationButtonPressed(String buttonId) {
 
     }
 
+    @Override
+    public String getCurrentComponentName() {
```
Comment (jinshin1013) [inline lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/TitleBarReactViewController.java]: It was complaining that `getCurrentComponetName` needs to be implemented. Should this return `null` instead?

File: lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/externalcomponent/ExternalComponentViewController.java [hunk 3]
```diff
@@ -87,6 +87,11 @@ public void applyBottomInset() {
         if (view != null) presenter.applyBottomInset(view, getBottomInset());
     }
 
+    @Override
+    public String getCurrentComponentName() {
```
Comment (jinshin1013) [inline lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/externalcomponent/ExternalComponentViewController.java]: You mentioned that external component won't have component name. So should this return `null`?

**Comment by jinshin1013** [comment]
@guyca I've tried to implement the Android part but having no prior exposure to Java made it really difficult. I've tried to follow your instruction as much as possible but Java was complaining about few things. I've got the tests to pass on Android and I've briefly tested on the Playground app and it successfully returned the `componentName` when dismissing modals.

iOS tests are failing with `The sandbox is not in sync with the Podfile.lock.`, I'll look into this soon.

**Comment by jinshin1013** [comment]
iOS E2E test failure doesn't seem to be related to this PR. The test is passing on my local machine but it isn't on the Jenkins.

File: lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/TitleBarButtonController.java (hunk 1 above)
Comment (guyca) [reply lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/TitleBarButtonController.java]: Yes

File: lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/TitleBarReactViewController.java (hunk 2 above)
Comment (guyca) [reply lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/TitleBarReactViewController.java]: Yes

File: lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/externalcomponent/ExternalComponentViewController.java (hunk 3 above)
Comment (guyca) [reply lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/externalcomponent/ExternalComponentViewController.java]: Return `externalComponent.name.get()` which is the name used to register the externalComponent with

**Comment by guyca** [comment]
Hey @jinshin1013, Sorry for the delay. I commented, let me know if you have more questions :)

File: lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/TitleBarButtonController.java (hunk 1 above)
Comment (jinshin1013) [reply lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/TitleBarButtonController.java]: Now returning:

```java
@Override
public String getCurrentComponentName() {
  return button.component.name.get();
}
```

File: lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/TitleBarReactViewController.java (hunk 2 above)
Comment (jinshin1013) [reply lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/TitleBarReactViewController.java]: Now returning:

```java
@Override
public String getCurrentComponentName() {
  return null;
}
```

File: lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/externalcomponent/ExternalComponentViewController.java (hunk 3 above)
Comment (jinshin1013) [reply lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/externalcomponent/ExternalComponentViewController.java]: Now returning:

```java
@Override
public String getCurrentComponentName() {
  return externalComponent.name.get();
}
```

**Comment by jinshin1013** [comment]
@guyca Thanks for the feedback! Could you leave the PR when you get a chance please?

File: docs/docs/events.md [hunk 4]
```diff
@@ -133,17 +133,45 @@ commandCompletedListener.remove();
...
 
 });
 ...
 // Unsubscribe
 modalDismissedListener.remove();
 ```
+|       Parameter         | Description |
+|:--------------------:|:-----|
+|**componentId**| Id of the disappearing component|
+|**componentName**|Registered name used when registering the component with `Navigation.registerComponent()`|
```
Comment (guyca) [inline docs/docs/events.md]: Lets change the description to "Registered name of the disappearing component". WDYT?

File: lib/src/interfaces/ComponentEvents.ts [hunk 5]
```diff
@@ -20,7 +20,7 @@ export interface NavigationButtonPressedEvent extends ComponentEvent {
 }
 
 export interface ModalDismissedEvent extends ComponentEvent {
-  componentId: string;
+  componentName?: string;
```
Comment (guyca) [inline lib/src/interfaces/ComponentEvents.ts]: Why is componentName optional? Seems like we're always emitting it

File: docs/docs/events.md [hunk 6]
```diff
@@ -133,17 +133,45 @@ commandCompletedListener.remove();
 |**completionTime**|Timestamp when the command, and consecutive animations, completed.|
 
 ## registerModalDismissedListener
-Invoked when modal dismissed.
+
+Called each time this component (Modal) is dismissed.
```
Comment (guyca) [inline docs/docs/events.md]: I find this description a bit misleading as a modal can be dismissed only once. This method is invoked each time a modal (any modal) is dismissed. I think we should keep the current description as the new description is misleading

Comment (jinshin1013) [reply docs/docs/events.md]: Done!

File: docs/docs/events.md (hunk 4 above)
Comment (jinshin1013) [reply docs/docs/events.md]: Oops. Bad copy and paste skill :( Fixed.

File: lib/src/interfaces/ComponentEvents.ts (hunk 5 above)
Comment (jinshin1013) [reply lib/src/interfaces/ComponentEvents.ts]: `dismissAllModals` do not emit `componentName`. I'm not sure whether it should return the name of the top modal component name? Let me know I can have a look.

Comment (guyca) [reply lib/src/interfaces/ComponentEvents.ts]: [You can get the componentName of the top modal like we get the id of the top modal](https://github.com/wix/react-native-navigation/blob/master/lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/modal/ModalStack.java#L84)

**Comment by guyca** [comment]
Hey @jinshin1013 
Seems like `push a screen to ensure its not there after reload` e2e test fails. Perhaps componentName is nil and you're trying to add it to a dictionary.

**Comment by jinshin1013** [comment]
@guyca Hey yeah I am trying to make sure a parity between iOS and Android so `dismissAllModals` returns `componentName` on iOS too. [Look here](https://github.com/wix/react-native-navigation/pull/5846/files#diff-402a1f2cdd98bfe5ae49b8eecb54f391R381). 

If there is only one modal (stack modal), `UIViewController* lastViewController` is `RNNStackModal` which does not seem to have `name` attribute. 

I can have `dismissAllModals` on iOS to not return `componentName` would there be another way?

**Comment by jinshin1013** [comment]
@guyca I know you are busy preparing V5 but could I get some guidance regarding the problem I mentioned above? Thanks 🙏

**Comment by jinshin1013** [comment]
Alright, I've tried to rebase the branch with the updated master and seems like I screwed up something. I'm going to close this one and open a new one.

**Comment by yogevbd** [comment]
@jinshin1013 Fixed the modals issue, can you please check? 😊

**Comment by jinshin1013** [comment]
@yogevbd Legend! Tested on the Playground, `dismissAllModals` do return `componentName` successfully. Thanks for that fix man :) Will keep my eye on the CI tests and hopefully this PR can get merged soon.

**Pull request merged**